File: z3wire/weave/emit_proto.py
"""Generates a .proto file from a resolved RDL module."""

from z3wire.weave.resolver import ResolvedModule


def _proto_type(field) -> str:
    """Map a resolved field to its proto type."""
    if field.kind == "bool":
        return "bool"
    elif field.kind == "bitvec":
        if field.signedness == "signed":
            return "sint64" if field.element_width > 32 else "sint32"
        return "uint64" if field.element_width > 32 else "uint32"
    elif field.kind == "enum_ref":
        # Enum fields use the underlying integer type
        return "uint64" if field.element_width > 32 else "uint32"
    elif field.kind == "struct_ref":
        return f"{field.struct_name}Proto"
    else:
        raise ValueError(f"Unknown field kind: {field.kind}")


def emit_proto(module: ResolvedModule) -> str:
    """Generate a .proto file string from a resolved module."""
    lines = []
    lines.append('syntax = "proto3";')
    lines.append(f"package {module.namespace};")
    lines.append("")

    for struct in module.structs:
        if struct.desc:
            lines.append(f"// {struct.desc}")
        lines.append(f"message {struct.name}Proto {{")
        field_number = 1
        for field in struct.fields:
            if field.reserved:
                continue
            proto_type = _proto_type(field)
            if field.count >= 1:
                lines.append(f"  repeated {proto_type} {field.name} = {field_number};")
            else:
                lines.append(f"  {proto_type} {field.name} = {field_number};")
            field_number += 1
        lines.append("}")
        lines.append("")

    return "\n".join(lines)

File: z3wire_weave/emit_proto.cc
#include "z3wire_weave/emit_proto.h"

#include <stdexcept>
#include <string>

#include "absl/strings/str_cat.h"
#include "z3wire_weave/resolver.h"

namespace z3wire_weave {

namespace {

std::string ProtoType(const ResolvedField& field) {
  switch (field.kind) {
    case ResolvedField::kBool:
      return "bool";
    case ResolvedField::kBitVec:
      if (field.signedness == ResolvedField::kSigned) {
        return field.element_width > 32 ? "sint64" : "sint32";
      }
      return field.element_width > 32 ? "uint64" : "uint32";
    case ResolvedField::kEnumRef:
      return field.element_width > 32 ? "uint64" : "uint32";
    case ResolvedField::kStructRef:
      return field.struct_name + "Proto";
    default:
      throw std::invalid_argument("Unknown field kind");
  }
}

}  // namespace

std::string EmitProto(const ResolvedModule& module) {
  std::string out;
  absl::StrAppend(&out, "syntax = \"proto3\";\n");
  absl::StrAppend(&out, "package ", module.ns, ";\n");
  absl::StrAppend(&out, "\n");

  for (const auto& s : module.structs) {
    if (!s.desc.empty()) {
      absl::StrAppend(&out, "// ", s.desc, "\n");
    }
    absl::StrAppend(&out, "message ", s.name, "Proto {\n");
    int field_number = 1;
    for (const auto& f : s.fields) {
      if (f.reserved) {
        continue;
      }
      std::string proto_type = ProtoType(f);
      if (f.count >= 1) {
        absl::StrAppend(&out, "  repeated ", proto_type, " ", f.name, " = ",
                        field_number, ";\n");
      } else {
        absl::StrAppend(&out, "  ", proto_type, " ", f.name, " = ",
                        field_number, ";\n");
      }
      ++field_number;
    }
    absl::StrAppend(&out, "}\n");
    absl::StrAppend(&out, "\n");
  }

  // Remove trailing blank line to match Python output format.
  if (!out.empty() && out.back() == '\n' && out.size() >= 2 &&
      out[out.size() - 2] == '\n') {
    out.pop_back();
  }

  return out;
}

}  // namespace z3wire_weave
